Prerelease1 start with parameters

tglomb opened the pull request:
The script 'scrnaseq.Rmd' has been adopted to receive parameter settings from outside including two small changes in functions_util.R. An R script for using the report from the command line is also included.

Commit:
Merge pull request #1 from MHH-RCUG/startScriptWithParametersFromOutsideForHTO

Start script with parameters from outside for hto

Commit:
Merge pull request #2 from ktrns/master

Update of MHH-RCUG/scrnaseq (master) with ktrns/scrnaseq (master)

Commit:
Merge pull request #3 from ktrns/master

Get latest version of scrnaseq

Commit:
Reintegration of violin plots
- Violin plots added
- figure height of DEG plots doubled
- xlab parameter added to PlotMyStyle function to be able to change xlab in violin plots

Commit:
Merge pull request #4 from MHH-RCUG/reintregrateViolinPlots

Reintegration of violin plots

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge pull request #6 from ktrns/master

Update of MHH-RCUG/scrnaseq (master) with ktrns/scrnaseq (master)

Commit:
Merge pull request #7 from ktrns/master

Update to pull last changes.

Commit:
Merge pull request #8 from ktrns/master

get the latest version

Commit:
Delete functions_plotting.R

Delete to pull new version from Dresden

Commit:
Create functions_plotting.R

Commit:
Merge pull request #10 from ktrns/prerelease1_andpet

Prerelease1 andpet

Commit:
Adopted script to receive parameter settings from outside.

Commit:
Update scrnaseq.Rmd

Some minor changes to get the script working with parameters from outside.

Commit:
Update functions_util.R

A small change to make the script working with parameters from outside.

Commit:
Create run_scrnaseq.R

Script to setup parameters and start rendering scrnaseq.Rmd from within R/RStudio.
To test the report stepwise in RStudio just use the setup part for the parameters and execute params=paramsList() before going through the chunks.

ktrns commented on R/functions_util.R:
```diff
@@ -427,7 +427,7 @@ check_parameters_scrnaseq = function(param) {
   
   # Check downsample_cells_n ###
-  if ("downsample_cells_n" %in% names(param)) {
+  if (("downsample_cells_n" %in% names(param)) & (!is.null(param$downsample_cells_n))) {
```
Andreas changed this part in his PR as well. So let's see after we merged his changes.

ktrns commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,58 @@
+paramsList = list()
```
Yes and no. Can you have a look at Andreas script run_scrnaseq_hto.R? It is written in a way that you can start it directly from command-line. That is in our case, we start the HTO script through a our Cell Ranger snakemake pipeline. 

Would you mind adapting your script to the style of the hto script? That would be super helpful.

ktrns commented on scrnaseq.Rmd:
````diff
@@ -159,189 +331,13 @@ invisible(knitcitations::citep(citation("knitr")))
-```
-
 ```{r preparation_and_initial_checks}
+# Copy input parameters to internal parameter list
````
I'd have to have a closer look but doesn't `functions_util.R` take care of this?

For example: 
```
  # Check samples_to_drop
  if ("samples_to_drop" %in% names(param) && !is.character(param$samples_to_drop)) {
    error_messages = c(error_messages, "The parameter 'samples_to_drop' is not a character vector. Please set to NULL if there are no samples to drop!")
  }
```

ktrns reviewed (commented):
Dear @tglomb, 

Thanks a lot for your PR. It is great to see that things are moving and we are getting closer to our release! 

My biggest request concerns `run_scrnaseq.R`. It would be very nice to have it in a generic way such that we can run it from command line, like we have it in place for the Pre-Workflow. Would you be able to implement this?

Commit:
Update functions_util.R

Correction of deg_contrasts check. Excel file could not be used before.

Commit:
Update run_scrnaseq.R

Script to setup parameters and start rendering scrnaseq.Rmd from within R/RStudio.
To test the report stepwise in RStudio just use the setup part for the parameters and execute params=paramsList() before going through the chunks.

Commit:
Add files via upload

Example files for start script.

tglomb commented:
> 
> 
> Dear @tglomb,
> 
> Thanks a lot for your PR. It is great to see that things are moving and we are getting closer to our release!
> 
> My biggest request concerns `run_scrnaseq.R`. It would be very nice to have it in a generic way such that we can run it from command line, like we have it in place for the Pre-Workflow. Would you be able to implement this?

Dear @ktrns,  
as previously anounced I have finally included the start script to be used from the command line. I also included two configuration files in the test directory (./test_datasets/10x_SmartSeq2_pbmc_GSE132044) to test the report with the settings from the original report.  
What did not work for me was to start knitting the report from the command line using relative paths. I had to provide the full paths to the start script. Maybe you could test it on your side.

ktrns commented:
Dear @tglomb, 

Is your PR complete, and ready for us to test? 

Thanks
Katrin

tglomb commented:
Dear @ktrns,  

yes, it's ready from my part.  
Happy testing! ... and success!

Kind regards  
Torsten

andpet0101 commented on R/functions_util.R:
```diff
@@ -756,7 +756,7 @@ check_parameters_scrnaseq = function(param) {
     }
     
-    if (!is.data.frame(param$deg_contrasts)) {
+    if (!is.data.frame(deg_contrasts_table)) {
```
Do you need the option to read contrasts from file? If not, we remove it

tglomb replied:
Yes, it's necessary for variable input to start the script automatically from command line. Please keep this.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+# Rscript scripts/run_scrnaseq.R --project-id=pbmc --path-data-csv=test_datasets/10x_SmartSeq2_pbmc_GSE132044/path_data.csv --path-out=test_datasets/10x_SmartSeq2_pbmc_GSE132044/results --deg-contrasts=test_datasets/10x_SmartSeq2_pbmc_GSE132044/deg_contrasts.xlsx --path-to-git=.
+#!/usr/bin/env Rscript
+
+getwd()
```
Is this necessary?

tglomb replied:
No, don't think so. Have it already removed.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+  epilog="Upon successful completion, the output directory will contain an HTML report, ....")
+
+# Set mutually exclusive groups
+normalizationGroup = parser$add_mutually_exclusive_group(required=FALSE)
```
Can we have a different name here since this is not just about normalisation? `analysisGroup` ?

tglomb replied:
That's no problem. I changed the name.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+speciesGroup = parser$add_mutually_exclusive_group(required=FALSE)
+speciesGroup$add_argument("--human", action="store_true", help="Annotation for human", required=FALSE)
+speciesGroup$add_argument("--mouse", action="store_true", help="Annotation for mouse", required=FALSE)
```
Can you add zebrafish?

```
  param[["mart_dataset"]] = "drerio_gene_ensembl"
  param[["annot_version"]] = 103
  param[["annot_main"]] = c(ensembl="ensembl_gene_id", symbol="external_gene_name", entrez="external_gene_name")
  param[["mart_attributes"]] = c(param$annot_main, c("chromosome_name", "start_position", "end_position",
                                                     "percentage_gene_gc_content", "gene_biotype", "strand", "description"))
  param[["mt"]] = "^mt-"
  param[["enrichr_dbs"]] = c("GO_Biological_Process_2018", "GO_Molecular_Function_2018", GO_Cellular_Component_2018)
  ```

tglomb replied:
Zebrafish has been added.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+  param[["mart_attributes"]] = c(param$annot_main, c("chromosome_name", "start_position", "end_position",
+                                                     "percentage_gene_gc_content", "gene_biotype", "strand", "description"))
+  param[["mt"]] = "^MT-"
+  param[["enrichr_dbs"]] = c("GO_Biological_Process_2018", "KEGG_2019_Human", "WikiPathways_2019_Human")
```
Could you add GO_MF and GO_CC as well for human and mouse?

tglomb replied:
Added both to human and mouse but left the others as it was. Is that correct?

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+# Get project and sample description
+parser$add_argument("--project-id", action="store", help="Project ID", dest="project_id", required=TRUE)
+
+parser$add_argument("--path-data-csv", action="store", help="CSV file containing the path data of the samples", dest="path_data_csv",required=TRUE)
```
Is it a comma-separated or a tab-separated file?

tglomb replied:
It's csv not tsv.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+# Get individual parameters
+parser$add_argument("--file-known-markers", action="store", help="Path to a xlsx file with marker genes based on literature, translated to Ensembl IDs (one list per column, first row as header and Ensembl IDs below); #default if no known marker genes should be plotted", dest="file_known_markers", default=NULL)
+
+parser$add_argument("--mart-dataset", action="store", help="Annotation via biomaRt; Dataset (default: #default)", dest="mart_dataset",default="hsapiens_gene_ensembl")
```
Can we make this a required parameter and not set a default?

tglomb replied:
Done.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+parser$add_argument("--file-known-markers", action="store", help="Path to a xlsx file with marker genes based on literature, translated to Ensembl IDs (one list per column, first row as header and Ensembl IDs below); #default if no known marker genes should be plotted", dest="file_known_markers", default=NULL)
+
+parser$add_argument("--mart-dataset", action="store", help="Annotation via biomaRt; Dataset (default: #default)", dest="mart_dataset",default="hsapiens_gene_ensembl")
+parser$add_argument("--annot-version", action="store", help="Annotation via biomaRt; Version (default: #default)", dest="annot_version",type="integer",default=98)
```
Same here?

tglomb replied:
Also done.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+parser$add_argument("--mart-attributes", action="store", help="Annotation via biomaRt; Attributes (default: #default)", dest="mart_attributes",default="chromosome_name,start_position,end_position,percentage_gene_gc_content,gene_biotype,strand,description")
+parser$add_argument("--biomart-mirror", action="store", help="Annotation via biomaRt; Mirror (default: #default)", dest="biomart_mirror", default=NULL)
+
+parser$add_argument("--file-annot", action="store", help="Alternatively, read a previously compiled annotation table from file (default: #default)", dest="file_annot", default=NULL)
```
FYI: We removed file-annot and we will take care of this here as well when merging.

tglomb replied:
Okay, great.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+parser$add_argument("--file-annot", action="store", help="Alternatively, read a previously compiled annotation table from file (default: #default)", dest="file_annot", default=NULL)
+
+parser$add_argument("--mt", action="store", help="Prefix of mitochondrial genes (default: #default)", dest="mt", default="^MT-")
```
Can we make this required? I think that you can nicely give default via your analysisGroup settings parameter, right?

tglomb replied:
Which parameter exactly do you mean - file-annot or mt?  
We can make it required but setting it later somewhere else would overwrite it and be kind of confusing. Are you sure you mean analysisGroup or the settings for a certain species?

ktrns replied:
@andpet0101 Can you double-check the reply?

ktrns replied:
Andreas meant `mt` ... Perhaps it is easier that we introduce this change after we did a merge.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+parser$add_argument("--mt", action="store", help="Prefix of mitochondrial genes (default: #default)", dest="mt", default="^MT-")
+
+parser$add_argument("--cell-filter-nFeature-RNA", action="store", help="Filter cells for nFeature_RNA (lower and upper threshold seperated by ',') (default: #default)", dest="cell_filter_nFeature_RNA", default="200, NA")
```
Here it would be nice if we can provide a more flexible option. How about this: 

- `--cell-filter cellMetadataColumn=min,max` (for numeric filters)
- `--cell-filter cellMetadataColumn=item1,item2,item3,..` (for level/character filters)

This option could be used multiple times...

ktrns replied:
@tglomb Is there anything to reply to this comment?

ktrns replied:
We'll take care of it.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+parser$add_argument("--cell-filter-nFeature-RNA", action="store", help="Filter cells for nFeature_RNA (lower and upper threshold seperated by ',') (default: #default)", dest="cell_filter_nFeature_RNA", default="200, NA")
+parser$add_argument("--cell-filter-percent-mt", action="store", help="Filter cells for percent_mt (lower and upper threshold seperated by ',') (default: #default)", dest="cell_filter_percent_mt", default="NA, 25")
+
+parser$add_argument("--feature-filter-min-counts", action="store", help="Filter features for min_counts; Minimum count of one feature (default: #default)", type="integer", dest="feature_filter_min_counts", default=1)
```
Same here?

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+parser$add_argument("--latent-vars", action="store", help="Additional (unwanted) variables to account for in statistical tests (default: #default)", dest="latent_vars", default=NULL)
+
+parser$add_argument("--deg-contrasts", action="store", help="Contrasts to find differentially expressed genes; xlsx file (Required columns: condition_column, condition_group1, condition_group2; Optional columns: subset_column, subset_group, assay, slot, padj, log2FC, min_pct, test, downsample_cells_n, latent_vars)", dest="deg_contrasts", required=TRUE)
```
Often we do not do DEG comparisons so can we make this optional (require=FALSE)?

tglomb replied:
Done.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+param[["annot_version"]] = opt[["annot_version"]]
+
+# Param annot_main
+param[["annot_main"]] = trimws(unlist(strsplit(trimws(unlist(strsplit(opt[["annot_main"]], ","))), "=")))[c(F,T)]
```
Can you please use the magrittr operator so that it is better readable?

```
opt["annot_main"] %>% 
  strsplit(",") %>%
  unlist() %>% ...
```

tglomb replied:
Changed the script accordingly and tried to use as much %>% as possible.

andpet0101 commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+parser$add_argument("--path-data-csv", action="store", help="CSV file containing the path data of the samples", dest="path_data_csv",required=TRUE)
+
+parser$add_argument("--downsample-cells-n", action="store", type="integer", help="Downsample data to at most N cells (default: #default)", dest="downsample_cells_n", default=0)
```
Default should be NULL.

tglomb replied:
Set the default to NULL but could have implications in the script if variable doesn't exist. Could you please test and see if it works?

ktrns replied:
@tglomb It currently is NULL if no downsampling is required. Hence that should work.

andpet0101 commented on test_datasets/10x_SmartSeq2_pbmc_GSE132044/path_data.csv:
```diff
@@ -0,0 +1,3 @@
+name,type,path,stats
```
Can you create this file in the `download.R` file so that paths are not hardcoded?

tglomb replied:
Created path_data.csv and deg_contrasts.xlsx in ```download.R```.

andpet0101 reviewed (commented):
We had a first look at the code and have written some first comments. We will now test the code.

ktrns commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+parser$add_argument("--cc-rescore-after-merge", action="store_true", help="Re-score cell cycle effects after data from different samples have been merged/integrated (default: #default)", dest="cc_rescore_after_merge", default=TRUE)
+parser$add_argument("--vars-to-regress", action="store", help="Additional (unwanted) variables that will be regressed out for visualisation and clustering (default: #default)", dest="vars_to_regress", default=NULL)
+
+parser$add_argument("--integrate-samples-method", action="store", help="Single or multiple datasets and how to combine them (method='single','merge', or 'integrate'); 'single': Default when there is only one dataset after filtering, no integration is needed. 'merge': Merge (in other words, concatenate) data when no integration is needed, e.g. when samples were multiplexed on the same chip. 'integ ...
```
I'd prefer these variables to be called `--combine-samples-method` and `--integrate-samples-options`. Ok?

tglomb replied:
The script has been adopted to style guide as much as possible or reasonable.  

The two variable names have been changed.

ktrns commented on scripts/run_scrnaseq.R:
```diff
@@ -0,0 +1,382 @@
+
+parser$add_argument("--marker-padj", action="store", type="double", help="Thresholds to define marker genes; Adjusted p-value (default: #default)", dest="marker_padj", default=0.05)
+parser$add_argument("--marker-log2FC", action="store", type="double", help="Thresholds to define marker genes; Fold change (default: #default)", dest="marker_log2FC", default=log2(2))
+parser$add_argument("--marker-pct", action="store", type="double", help="Thresholds to define marker genes; Percentage of marker expressing cells (default: #default)", dest="marker_pct", default=0.25)
```
The default seems a bit strict. Seurat asks for 0.1 in `FindAllMarkers`.

tglomb replied:
I can change it to 0.1 but 0.25 is the value set in scrnaseq.Rmd. Just tell me please what you prefer.

ktrns reviewed (changes requested):
Hi Torsten. 

Thanks again for your work. 

I had these two additional comments. Plus a general favour to ask you: Could you adopt the style we have in the style guide? That is, preferably use `%>%` over multiple `((((`, always add a space after a comma, FALSE/TRUE rather than F/T. 

We are now looking into what the best strategy is to test your branch, as we need to find a way to merge the new master first, at least for testing.

andpet0101 commented:
I have downloaded your PR and tried to merge the current master into it. There were two conflicts in `scrnaseq.Rmd` and one in `R/functions_util.R`. Can you have a look at them?
- The first one in `scrnaseq.Rmd` is probably because because we added a `param$assay_raw = "RNA"` which contains the name of raw data assay. Could you simply add it as well?
- The second one in `scrnaseq.Rmd` is likely caused by the additional params section in the YAML header. Here your version should be fine.
- The one in `R/functions_util.R`: Here the master version is probably better since the `&` is used as a short-cut and version. If the first part is already false, the second part will not be evaluated. With `&&`, both parts are always evaluated (which could cause a name error/not in list error).

Commit:
Update run_scrnaseq.R

Commit:
Update run_scrnaseq.R

Changed normalizationGroup to analysisGroup

Commit:
Update run_scrnaseq.R

Set mart-dataset as required

Commit:
Update run_scrnaseq.R

Set annot-version as required.

Commit:
Update run_scrnaseq.R

Added zebrafish as species.

Commit:
Update run_scrnaseq.R

Added "GO_Molecular_Function_2018" and "GO_Cellular_Component_2018" for human and mouse.

Commit:
Update run_scrnaseq.R

Parameter deg-contrasts made unrequired.

Commit:
Update run_scrnaseq.R

Set default of downsample-cells-n to NULL.

Commit:
Update download.R

Added output of path_data.csv.

Commit:
Delete path_data.csv

Removed because it will be generated in download.R.

Commit:
Update download.R

Added generation of deg_contrasts.xlsx.

Commit:
Delete deg_contrasts.xlsx

Deleted because it will be generated in download.R.

Commit:
Update run_scrnaseq.R

Introduced magrittr to the script and made some changes respectively.

Commit:
Update run_scrnaseq.R

Did my best to adopt the style guide the script.

Commit:
Update run_scrnaseq.R

The two variables --integrate-samples-method and --integrate-samples-integrate were changed to --combine-samples-method and --integrate-samples-options.

tglomb commented:
Hi @ktrns and @andpet0101,  

could you please have a look on the pull request again to see if I resolved all issues or if there is anything else to do.  

Best regards  
Torsten

ktrns reviewed (approved):
We'll try to merge your changes, and take care of the remaining small open questions.
